### DISCUSSION TITLE
SSR Usage - `useHydrateAtoms` hook causes a UI missmatch

### DISCUSSION THREAD

**Discussion by fabio-nettis** [pmndrs/jotai#1982] [Q&A] [answered]
Working on a Next.js 13 project using the new app directory and jotai as global state manager. Now I am trying to pass an initial state to my atoms using the `useHydrateAtoms` hook as seen in the [official documentation](https://jotai.org/docs/utilities/ssr), but this causes a hydration error.

Following is an simplified version of the code that is currently used, it passes the data received from the server side component to a client component that calls `useHydrateAtoms` and uses `useAtom` to read and write from/to that atom. 

##### Server component (page)

```tsx
const getData = async () => {
 // ...
}

export default async function Page() {
  const data = await getData();
  return <ChildComponent initialState={data} />;
}
```

##### Client component

```tsx
"use client";

export function ChildComponent({ initialState }) {
  useHydrateAtoms([[myAtom, initialState]]);
  const [data, setData] = useAtom(myAtom);
  return <>{data.id}</>;
}
```

When I dynamically import the child component using `next/dynamic` with the SSR option set to false or I remove the `useHydrateAtoms` hook, the error disappears completely but both of those solutions defy the purpose of this implementation. 

If not like this, how can one provide an initial state to an atom with a value from the server, so the atom is not `null` upon the first client side render?

**Comment by dai-shi** [comment]
> How can one provide an initial state to my atom with a value from the server so the atom is not null upon the first client side render?

Is it null? because, your usage seems correct.

Can anyone help on this?

**Reply by fabio-nettis** [reply]
It is correctly set. But is causes a UI missmatch as stated above.

**Reply by dai-shi** [reply]
> a UI missmatch

what is expected and what is actual?

**Reply by fabio-nettis** [reply]
Expected would be that the atom is populated and everything works fine, but actuall; While the atom is being populated a hydration error is thrown.

**Reply by dai-shi** [reply]
Does the error show what is the expected value?

**Reply by fabio-nettis** [reply]
The value is set. Yes. But as I said, a hydration error is thrown right after the atoms population.

**Reply by wootra** [reply]
is it resolved? @fabio-nettis if it is, can you share why it is closed? it will be a great opportunity for the jotai users.

**Reply by fabio-nettis** [reply, marked as answer]
@wootra  I have discovered that the `useHydrateAtoms` hook was not the issue per-se, but if understood correctly it was a client component accessing the state in a different part of the UI imported dynamically via `next/dynamic` with the `ssr` option set to false.  This caused the value displayed by the component to change on first render, what in return caused a UI mismatch.

#### How to I fixed the error

Simply move the component that calls useHydrateAtoms to the top of the component tree, this will ensure that all children will render with the correct value.
